### PR TITLE
do not use lock for async SGD in ReportGradient

### LINKED ISSUE #1167: Do not use lock in ReportGradient if async

### PR DESCRIPTION
**Pull request by skydoorkai**
fix #1167

### CODE REVIEW

File: elasticdl/python/master/servicer.py [hunk 1]
```diff
@@ -229,11 +223,19 @@ def _update_model(self):
...
 
+        # need the lock for model version update in async SGD
+        if self._use_async:
+            self._lock.acquire()
         self._update_model_version()
-        self._gradient_sum.clear()
-        self._gradient_sum_indexed.clear()
-        self._edl_embedding_gradients.clear()
-        self._grad_n = 0
+        self._update_evaluation()
```
Comment (chunyang-wen) [inline elasticdl/python/master/servicer.py:230]: Can we wrap these clauses into a `with` statemennt?

File: elasticdl/python/master/servicer.py [hunk 2]
```diff
@@ -401,16 +399,19 @@ def ReportGradient(self, request, _):
...
+            # TODO: do not accumulate gradients but apply directly.
+            pass
+
+        # staleness-aware learning rate modulation
+        if self._lr_modulation:
+            staleness = max(1, self._version - request.model_version)
+            self._lr_modulation.set_multiplier(1.0 / staleness)
+        if self._use_async or self._grad_n >= self._grad_to_wait:
+            self._update_model()
+        if not self._use_async:
```
Comment (chunyang-wen) [inline elasticdl/python/master/servicer.py]: `lock.acquire` and `lock.release` appears in multiple places. It is easy to mess things up.

File: elasticdl/python/master/servicer.py (hunk 1 above)
Comment (ywskycn) [reply elasticdl/python/master/servicer.py:230]: May not easy to put `with` here, as the lock is only for `_use_async`.

File: elasticdl/python/master/servicer.py (hunk 2 above)
Comment (ywskycn) [reply elasticdl/python/master/servicer.py]: One possible idea is to move some intermediate code outside to another func to make the acquire->func->release clearer.

Comment (skydoorkai) [reply elasticdl/python/master/servicer.py]: changed

File: elasticdl/python/master/servicer.py [hunk 3]
```diff
@@ -229,11 +223,19 @@ def _update_model(self):
...
-        self._gradient_sum_indexed.clear()
-        self._edl_embedding_gradients.clear()
-        self._grad_n = 0
+        self._update_evaluation()
+        self._update_checkpoint()
+        if self._use_async:
+            self._lock.release()
+        else:
+            self._gradient_sum.clear()
+            self._gradient_sum_indexed.clear()
```
Comment (chunyang-wen) [inline elasticdl/python/master/servicer.py:236]:
```python
from contextlib import contextmanager
@contextmanager
def _possible_lock():
    if self._use_async:
        self._lock.acquire()

     yield
     if self._use_async:
         self._lock.release()

with _possible_lock():
  self._update_model_version()
  self._update_evaluation()
  self._update_checkpoint()

if not self._use_async:
  self._graidentxx
```

File: elasticdl/python/master/servicer.py [hunk 4]
```diff
@@ -324,58 +326,68 @@ def ReportGradient(self, request, _):
...
+            else:
+                if arr.shape != self._model[k].numpy().shape:
+                    raise ValueError(
+                        "Gradient key: %s has incompatible dimension", k
+                    )
+                tmp[k] = arr
+
+        if not self._use_async:
+            self._lock.acquire()
+        self._process_gradients(
```
Comment (chunyang-wen) [inline elasticdl/python/master/servicer.py:377]: We can use a similar context as above comment.

**Review by chunyang-wen — APPROVED** [review]

**Pull request merged**